orion-lib/src/listeners/listeners_manager.rs

```diff
@@ -125,35 +155,94 @@ impl ListenersManager {
+
+        info!("Starting new version {} of listener {}", version, listener_name);
+
+        let listener_name_clone = listener_name.clone();
+
         let join_handle = tokio::spawn(async move {
             let error = listener.start().await;
-            warn!("Listener {listener_name} exited: {error}");
+            warn!("Listener {} version {} exited: {}", listener_name_clone, version, error);
         });
-        #[cfg(debug_assertions)]
-        if self.listener_handles.contains_key(&listener_name) {
-            debug!("Listener {listener_name} already exists, replacing it");
-        }
-        // note: join handle gets overwritten here if it already exists.
-        // handles are abort on drop so will be aborted, closing the socket
-        // but the any tasks spawned within this task, which happens on a per-connection basis,
-        // will survive past this point and only get dropped when their session ends
-        self.listener_handles.insert(listener_name, ListenerInfo::new(join_handle, listener_conf));
+
+        let listener_info = ListenerInfo::new(join_handle, listener_conf, version);
+
+        let versions = self.listener_handles.entry(listener_name.clone()).or_default();
+        versions.push(listener_info);
+        info!("Listener {} now has {} active version(s)", listener_name, versions.len());
+
+        self.cleanup_old_versions(&listener_name);
 
         Ok(())
     }
 
     pub fn stop_listener(&mut self, listener_name: &str) -> Result<()> {
-        if let Some(abort_handler) = self.listener_handles.remove(listener_name) {
-            info!("{listener_name} : Stopped");
-            abort_handler.handle.abort();
+        if let Some(listeners) = self.listener_handles.remove(listener_name) {
+            info!("Stopping all {} version(s) of listener {}", listeners.len(), listener_name);
+            for listener_info in listeners {
+                info!("Stopping listener {} version {}", listener_name, listener_info.version);
+                listener_info.handle.abort();
+            }
+        } else {
+            info!("No listeners found with name {}", listener_name);
         }
 
         Ok(())
     }
+
+    fn cleanup_old_versions(&mut self, listener_name: &str) {
+        if let Some(versions) = self.listener_handles.get_mut(listener_name) {
+            let original_count = versions.len();
+
+            match &self.config.cleanup_policy {
+                CleanupPolicy::CountBasedOnly(max_count) => {
+                    if versions.len() > *max_count {
+                        let to_remove = versions.len() - max_count;
+                        for _ in 0..to_remove {
+                            let old = versions.remove(0);
+                            info!("Cleaning up old listener {} version {} (count limit)", listener_name, old.version);
+                        }
+                    }
+                },
+                CleanupPolicy::TimeBasedOnly(_timeout) => {
```

Comment:
Keep it simple and remove if it is not implemented properly. Don't like that we will introduce potentially incorrect behaviour and rely on comments/TODOs to keep track of it.